gh-104050: Annotate Argument Clinic parameter permutation helpers

erlend-aasland opened the pull request:
<!-- gh-issue-number: gh-104050 -->
* Issue: gh-104050
<!-- /gh-issue-number -->

erlend-aasland committed:
gh-104050: Annotate Argument Clinic parameter permutation helpers

AlexWaygood commented on Tools/clinic/clinic.py:
```diff
@@ -28,11 +28,12 @@
 from typing import (
     Any,
     Final,
+    Generator,
```
Nit: `Generator` can also be imported from `collections.abc` rather than `typing` :)

erlend-aasland replied:
Yes! _That_ I actually knew 😄 Thanks!

erlend-aasland committed:
Address review: import Generator from collections.abc

AlexWaygood commented on Tools/clinic/clinic.py:
```diff
@@ -516,7 +516,13 @@ class PythonLanguage(Language):
+
+def permute_left_option_groups(
+        l: Reversible[ParamGroup]
+) -> ParamGenerator:
```
Several comments here:
1. I think you _might_ be overdoing it a bit with the type aliases from time to time -- here I think it maybe makes it harder to spot the bug in your annotation. You've annotated the function as returning `Generator[tuple[Iterable[Parameter], ...], None, None]`. A function that did that would look something like this:

   ```py
   def foo():
       while True:
           yield ([parameter], [parameter], [parameter])
   ```

   But I don't think that's what this function does. ISTM that you _give_ this function `Iterable[Iterable[X]]` and it returns a generator that yields `tuple[X, ...]`. So that would imply that the type hint should be `Generator[tuple[Parameter, ...], None, None]`
2. Having just told you to use `collections.abc.Generator` instead of `typing.Generator`... I actually wouldn't use `Generator` at all here, now that I take a closer look! The fact that this function is a generator function is an implementation detail that could easily go away in a later refactor. You're never `send`ing any values _into_ the generator, only getting values yielded _from_ the generator, so it's indistinguishable from any other iterator. As such, I'd go with a simpler and more readable type hint here for the return type -- either `Iterator[tuple[Parameter, ...]]` or `Iterable[tuple[Parameter, ...]]`. (I lean towards `Iterator` for generator-function return types, but there's not much difference between them.)
3. You've annotated the parameter with `Reversible[Iterable[Parameter]]`. `Reversible[X]` means "an object which, when `reversed()` is called on it, turns into `X`". So `Reversible[Iterable[Parameter]]` implies that the function would accept objects such as `[parameter, parameter, parameter]` -- when you call `reversed()` on a `list[parameter]`, you get an `Iterable[parameter]`, so this satisfies the annotation you've given. But that wouldn't be enough for this function -- this function requires an object which, when reversed, gives you an iterable of iterables of parameters. You _could_ annotate this with `Reversible[Iterable[Iterable[Parameter]]]`... or you could just do something more simple and readable (which should be fine for our purposes) -- `Sequence[Iterable[Parameter]]` :)

So, putting all of that together... here's how I'd annotate this function:

```py
def permute_left_option_groups(
    l: Sequence[Iterable[Parameter]]
) -> Iterator[tuple[Parameter, ...]]:
```

AlexWaygood replied:
(Note that I'm _not_ saying "don't use any type aliases" -- they can often greatly improve readability! But deeply nested type aliases can also _hurt_ readability by obscuring what the actual type is -- it can cut both ways; you need to take it on a case-by-case basis imo)

erlend-aasland replied:
In a previous version, I had four type aliases; that did not help 😆 Thanks for all the info! I'll have another go.

erlend-aasland committed:
Address review:

- Use a single alias: ParamTuple
- Ditch generators, use Iterator
- Use Sequence iso. layers of Iterable/Reversible/etc.

AlexWaygood commented on Tools/clinic/clinic.py:
```diff
@@ -540,13 +552,17 @@ def permute_right_option_groups(l):
+def permute_optional_groups(
+        left: Sequence[ParamTuple],
+        required: ParamTuple,
+        right: Sequence[ParamTuple]
```
Mandating that the first and third parameters have to be passed `Sequence[tuple[Parameter, ...]]` rather than `Sequence[Iterable[Parameter]]` makes the type hints simpler now, but it will cause problems later on! You'll notice that this function is only ever actually _called_ once in `clinic.py`, and the one time it _is_ called... it's passed a `list[list[Parameter]]` for the first and third parameter!

https://github.com/python/cpython/blob/eeb5c63179eba0390c8937d0803a41092225be7e/Tools/clinic/clinic.py#L1371

So, you'll either need to change the annotations here to something less specific (`Sequence[Iterable[Parameter]]`), or change the place where the function is called so that it's passed objects of `list[tuple[Parameter, ...]]` instead of objects of type `list[list[Parameter]]`

AlexWaygood replied:
(The reason why mypy isn't flagging this issue now is because we have our mypy config set up so that mypy will only check functions with type annotations, and the place where this function is called is in a function that doesn't have type annotations.)

erlend-aasland replied:
Darn, you're right. There is no easy way out of this.

AlexWaygood replied:
Yeah -- this is why I used different types for the return types and parameter types in the diff I suggested in https://github.com/python/cpython/pull/106431#issuecomment-1620696237 :)

erlend-aasland replied:
Ok, PTAL! I broke this up in two commits; I'm not sure you'll like the latter (8d1821c26aa29e4bc9a1a83bb853d83ba17e9920) :)

AlexWaygood replied:
Looks great! Hard to find perfect names for these type aliases, but I think the aliases themselves do make sense; I'm fine with 8d1821c :)

Just one nit: https://github.com/python/cpython/pull/106431#discussion_r1252375853

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Address review: make it less specific

erlend-aasland committed:
Add ParamGroup alias

AlexWaygood commented on Tools/clinic/clinic.py:
```diff
@@ -540,13 +553,17 @@ def permute_right_option_groups(l):
-def permute_optional_groups(left, required, right):
+def permute_optional_groups(
+        left: Sequence[ParamGroup],
+        required: Iterable[Parameter],
```
```suggestion
        required: ParamGroup,
```

erlend-aasland committed:
Update Tools/clinic/clinic.py

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

AlexWaygood reviewed (approved):
_Now_ we're there ;)

This was a very tricky one!
